### PR TITLE
WIP: Fix deprecations and some general improvements

### PR DESCRIPTION
**Pull request by sandreas**
Thanks for this awesome project, I'm impressed. Would love to see more community activity here... Tried to improve some minor things reported by `cargo check` and *IntelliJ* in this pull request. Since I'm pretty new to rust, i hope i did not break anything....

This project is nearly exactly what i was looking for. One thing i would like to implement is a recursive file iterator to enable standalone usage (no `rdup` required) considering https://github.com/sharkdp/fd as a start. I have to find out how `rdup` and `rdup-up` exactly work, but what do you think about a wrapper around `&mut io::stdin()` to provide a reader that merges/redirects a possible input argument iterator AND all `stdin`? Like:

```
# still works
rdup -x /dev/null "$HOME" | rdedup store home
rdedup load home | rdup-up "$HOME.restored"

# also works without rdup
rdedup store home "$HOME"
rdedup load home "$HOME.restored"
```

Before doing so i have to read some more code and further rust docs first - this multithreading code is far from being easy to understand - but maybe i do not need to even touch many parts of the code to get the above working.

### CODE REVIEW

**Comment by Ralith** [comment]
rdedup is designedto be usable as a library; an integrated front-end can therefore be an independent project. See [rackup](https://github.com/Ralith/rackup) for an (old, unmaintained) proof of concept.

**Comment by sandreas** [comment]
> rdedup is designedto be usable as a library; an integrated front-end can therefore be an independent project.
Thank you for pointing this out - i did not know the main goal was that.

> See rackup for an (old, unmaintained) proof of concept.
Nice to know but so i think i have some work to do. If you know a *ready to use* software that uses this lib, it would be great, if you tell me ;)

**Comment by dpc** [comment]
Hi. Sorry for lack of response. I'm quite busy. If I forget about you, please ping me again here on gitter, etc.

**Comment by dpc** [comment]
Can you bump the minimum version in `.travis.yaml` so the tests are passing again?

**Comment by dpc** [comment]
See https://github.com/dpc/rdedup/issues/5 for more chat about traversing dirs.

**Comment by dpc** [comment]
I actually have nothing against having a built-in dir traverser, as long as it is not impeding other possible implementations. And if it is to be built-in into `rdedup` it better be good. :)

**Comment by sandreas** [comment]
> I actually have nothing against having a built-in dir traverser, as long as it is not impeding other possible implementations. And if it is to be built-in into rdedup it better be good. :)

Thx for pointing out #5 - I would also take into account https://github.com/BurntSushi/walkdir, but perhaps i find something faster / better.

> Can you bump the minimum version in .travis.yaml so the tests are passing again?

Ok, bumped to latest (`1.39.0`).

Two other issues reported by my IDE - empty `matrix` value which i removed and invalid value for `on_start`. I hope this was not intended.

``` 
env:
  global:
    - RUST_BACKTRACE=1
    - RUST_TEST_THREADS=1
  matrix:
    -
      ^^^^---- something is missing here
    - RELEASE=true

...

notifications:
  webhooks:
    on_success: change  # options: [always|never|change] default: always
    on_failure: always  # options: [always|never|change] default: always
    on_start: false     # default: false
                   ^^^^^---- allowed values are always|never|change

```

I changed it to:

```
...
env:
  global:
    - RUST_BACKTRACE=1
    - RUST_TEST_THREADS=1
  matrix:
    - RELEASE=true
...

notifications:
  webhooks:
    on_success: change  # options: [always|never|change] default: always
    on_failure: always  # options: [always|never|change] default: always
    on_start: never     # default: false
```

File: .travis.yml [hunk 1]
```diff
@@ -24,7 +24,6 @@ env:
     - RUST_BACKTRACE=1
     - RUST_TEST_THREADS=1
   matrix:
-    -
```
Comment (dpc) [inline .travis.yml]: I thought this one is neccessary to have env without `RELEASE` set

File: .travis.yml [hunk 2]
```diff
@@ -39,4 +38,4 @@ notifications:
   webhooks:
     on_success: change  # options: [always|never|change] default: always
     on_failure: always  # options: [always|never|change] default: always
-    on_start: false     # default: false
```
Comment (dpc) [inline .travis.yml]: What's this for? :)

Comment (dpc) [reply .travis.yml]: Oh, OK.

File: .travis.yml (hunk 1 above)
Comment (dpc) [reply .travis.yml]: There's nothing wrong with empty yaml list elements, AFAIK. :)

Comment (sandreas) [reply .travis.yml]: > There's nothing wrong with empty yaml list elements, AFAIK.

Ok, if there is a reason, lets keep it. My IDE reported a problem here. I will fix this in the next commit.

File: .travis.yml (hunk 2 above)
Comment (sandreas) [reply .travis.yml]: > What's this for? :)

Well `false` seems to be an invalid value, especially because it is not the default (any more)...

**Comment by sandreas** [comment]
Ok, lets summarize the todos:

- fix .travis.yml by adding `-` to `matrix` again
- fix pending build issues (deprecations like `weak_rng`, `tester/src/bin.rs:51:11`, that were removed in later rust versions, see output below)

```
error[E0425]: cannot find function `weak_rng` in crate `rand`
  --> tester/src/bin.rs:51:11
   |
51 |     rand::weak_rng().gen_iter().take(len).collect::<Vec<u8>>()
   |           ^^^^^^^^ not found in `rand`
error[E0425]: cannot find function `sample` in crate `rand`
   --> tester/src/bin.rs:142:19
    |
142 |             rand::sample(&mut thread_rng(), self.names.keys(), 1)[0];
    |                   ^^^^^^ not found in `rand`
help: possible candidate is found in another module, you can import it into scope
    |
6   | use rand::seq::index::sample;
    |
error: aborting due to 2 previous errors
For more information about this error, try `rustc --explain E0425`.
error: could not compile `tester`.
To learn more, run the command again with --verbose.
/home/travis/.travis/functions: line 109: ./target/release/tester: No such file or directory
The command "if [ ! -z "$RELEASE" ] ; then cd tester; cargo build --release; cd ..; rm -Rf /tmp/rdedup-tester/; ./target/release/tester 200; fi" exited with 127.
591.99s$ if [ ! -z "$RELEASE" ] ; then ./scripts/e2e-test.sh; fi
```

I think i can fix these in the next days.

**Comment by sandreas** [comment]
Ok, fixed the error on rust (stable). I'm not sure, what the error on beta and nightly is, but i think it has nothing to do with the pull request... Perhaps it is because of the still existing deprecations?!

Is it ready for merging into master now or would you like to request some other changes?

**Comment by dpc** [comment]
Awesome! Let's land. No need to blow this PR too much.

**Pull request merged**